src/lib/pubkey/ec_group/ec_inner_pc.cpp

```diff
@@ -0,0 +1,209 @@
+*/
+
+#include <botan/internal/ec_inner_pc.h>
+
+namespace Botan {
+
+namespace {
+
+PCurve::PrimeOrderCurve::AffinePoint deserialize_pcurve_pt(const PCurve::PrimeOrderCurve& curve,
+                                                           std::span<const uint8_t> bytes) {
+   if(auto pt = curve.deserialize_point(bytes)) {
+      return *pt;
+   } else {
+      throw Decoding_Error("Invalid elliptic curve point encoding");
+   }
+}
+
+}  // namespace
+
+const EC_Scalar_Data_PC& EC_Scalar_Data_PC::checked_ref(const EC_Scalar_Data& data) {
+   const auto* p = dynamic_cast<const EC_Scalar_Data_PC*>(&data);
+   if(!p) {
+      throw Invalid_State("Failed conversion to EC_Scalar_Data_PC");
+   }
+   return *p;
+}
+
+const std::shared_ptr<const EC_Group_Data>& EC_Scalar_Data_PC::group() const {
+   return m_group;
+}
+
+size_t EC_Scalar_Data_PC::bytes() const {
+   return this->group()->order_bytes();
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::clone() const {
+   return std::make_unique<EC_Scalar_Data_PC>(this->group(), this->value());
+}
+
+bool EC_Scalar_Data_PC::is_zero() const {
+   return this->value().is_zero();
+}
+
+bool EC_Scalar_Data_PC::is_eq(const EC_Scalar_Data& other) const {
+   return (value() == checked_ref(other).value());
+}
+
+void EC_Scalar_Data_PC::assign(const EC_Scalar_Data& other) {
+   m_v = checked_ref(other).value();
+}
+
+void EC_Scalar_Data_PC::square_self() {
+   // TODO square in place
+   m_v = m_v.square();
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::negate() const {
+   return std::make_unique<EC_Scalar_Data_PC>(m_group, m_v.negate());
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::invert() const {
+   return std::make_unique<EC_Scalar_Data_PC>(m_group, m_v.invert());
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::add(const EC_Scalar_Data& other) const {
+   return std::make_unique<EC_Scalar_Data_PC>(m_group, m_v + checked_ref(other).value());
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::sub(const EC_Scalar_Data& other) const {
+   return std::make_unique<EC_Scalar_Data_PC>(m_group, m_v - checked_ref(other).value());
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::mul(const EC_Scalar_Data& other) const {
+   return std::make_unique<EC_Scalar_Data_PC>(m_group, m_v * checked_ref(other).value());
+}
+
+void EC_Scalar_Data_PC::serialize_to(std::span<uint8_t> bytes) const {
+   BOTAN_ARG_CHECK(bytes.size() == m_group->order_bytes(), "Invalid output length");
+   m_group->pcurve().serialize_scalar(bytes, m_v);
+}
+
+EC_AffinePoint_Data_PC::EC_AffinePoint_Data_PC(std::shared_ptr<const EC_Group_Data> group,
+                                               PCurve::PrimeOrderCurve::AffinePoint pt) :
+      m_group(std::move(group)),
+      m_pt(std::move(pt)),
+      m_bytes(m_pt.serialize<secure_vector<uint8_t>>()),
+      m_x_bytes(std::span{m_bytes}.subspan(1, field_element_bytes())),
+      m_y_bytes(std::span{m_bytes}.last(field_element_bytes())) {}
```

Comment:
Perhaps add a `BOTAN_ASSERT` (potentially `BOTAN_DEBUG_ASSERT`) to the constructor's body to check that `m_bytes` contains the expected number of bytes (`1+2*field_element_bytes()`). If, for any reason, `m_pt.serialize()` didn't produce the expected number of bytes, the spans into that buffer would potentially point to invalid memory. Same in the constructor below.

Reply:
Alternatively, one might explicitly allocate `m_bytes` in the constructor's body (as `m_bytes.resize(2*field_element_bytes()+1)`) and then use `m_pt.serialize_to(m_bytes)`. That would also make sure that the spans aren't pointing into the abyss.

Reply:
Added an assert here